core/shared/src/main/scala/zio/internal/impls/PartitionedRingBuffer.scala

```diff
@@ -0,0 +1,171 @@
+ */
+
+package zio.internal
+
+import zio.stacktracer.TracingImplicits.disableAutoTrace
+
+import java.util.concurrent.ThreadLocalRandom
+
+private[zio] final class PartitionedRingBuffer[A <: AnyRef](
+  preferredPartitions: Int,
+  preferredCapacity: Int,
+  roundToPow2: Boolean
+) extends MutableConcurrentQueue[A]
+    with Serializable {
+
+  private[this] val mask: Int    = MutableConcurrentQueue.roundToPow2MinusOne(preferredPartitions)
+  private[this] val nQueues: Int = mask + 1
+
+  private[this] val partitionSize: Int = {
+    val cap = Math.ceil(preferredCapacity.toDouble / nQueues).toInt.max(2)
+    if (roundToPow2) RingBuffer.nextPow2(cap) else cap
+  }
+
+  private[this] val queues = Array.fill(nQueues)(RingBuffer[A](partitionSize))
+
+  override final val capacity = nQueues * partitionSize
+
+  override def size(): Int = {
+    val from = ThreadLocalRandom.current().nextInt(nQueues)
+    var i    = 0
+    var size = 0
+    while (i < nQueues) {
+      val idx = (from + i) & mask
+      size += queues(idx).size()
+      i += 1
+    }
+    size
+  }
+
+  override def enqueuedCount(): Long = {
+    val from = ThreadLocalRandom.current().nextInt(nQueues)
+    var i    = 0
+    var size = 0L
+    while (i < nQueues) {
+      val idx = (from + i) & mask
+      size += queues(idx).enqueuedCount()
+      i += 1
+    }
+    size
+  }
+
+  override def dequeuedCount(): Long = {
+    val from = ThreadLocalRandom.current().nextInt(nQueues)
+    var i    = 0
+    var size = 0L
+    while (i < nQueues) {
+      val idx = (from + i) & mask
+      size += queues(idx).dequeuedCount()
+      i += 1
+    }
+    size
+  }
+
+  /**
+   * Offers an element to the queue, returning whether the offer was successful.
+   *
+   * @param maxMisses
+   *   How many partitions to try before giving up
+   */
+  def offer(a: A, random: ThreadLocalRandom, maxMisses: Int): Boolean = {
+    val maxI   = nQueues.min(maxMisses + 1)
+    val from   = random.nextInt(nQueues)
+    var i      = 0
+    var result = false
+    while (i < maxI && !result) {
+      val idx = (from + i) & mask
+      result = queues(idx).offer(a)
+      i += 1
+    }
+    result
+  }
+
+  def randomPartition(random: ThreadLocalRandom): RingBuffer[A] =
+    queues(random.nextInt(nQueues))
+
+  override def offer(a: A): Boolean =
+    offer(a, ThreadLocalRandom.current(), nQueues)
+
+  /**
+   * Polls an element to the queue, returning the element or `default` if the
+   * queue is empty.
+   *
+   * @param maxMisses
+   *   How many partitions to try polling from before giving up
+   */
+  def poll(default: A, random: ThreadLocalRandom, maxMisses: Int): A = {
+    val from   = random.nextInt(nQueues)
+    var i      = 0
+    var result = null.asInstanceOf[A]
+    var misses = 0
+    while ((result eq null) && i < nQueues && misses <= maxMisses) {
+      val idx   = (from + i) & mask
+      val queue = queues(idx)
+      result = queue.poll(default)
+      if (result eq null) misses += 1
```

Comment:
I think according to the method docs (and the semantics inherited from RingBuffer),
u're supposed to compare with default

Reply:
Ahh yeah nicely spotted!